api/metrics/v1/rules.go

```diff
@@ -33,8 +47,56 @@ func (rh *rulesHandler) get(w http.ResponseWriter, r *http.Request) {
 
 	defer resp.Body.Close()
 
-	if _, err := io.Copy(w, resp.Body); err != nil {
-		http.Error(w, "error writing rules response", http.StatusInternalServerError)
+	if resp.StatusCode/100 != 2 {
+		http.Error(w, "error listing rules", resp.StatusCode)
+		return
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		http.Error(w, "error listing rules", http.StatusInternalServerError)
+		return
+	}
+
+	var rawRules rules.Rules
+	if err := yaml.Unmarshal(body, &rawRules); err != nil {
+		level.Error(rh.logger).Log("msg", "could not unmarshal rules", "err", err.Error())
+		http.Error(w, "error unmarshaling rules", http.StatusInternalServerError)
+
+		return
+	}
+
+	for i := range rawRules.Groups {
+		for j := range rawRules.Groups[i].Rules {
+			switch r := rawRules.Groups[i].Rules[j].(type) {
+			case rules.RecordingRule:
+				if r.Labels.AdditionalProperties == nil {
+					r.Labels.AdditionalProperties = make(map[string]string)
+				}
+
+				r.Labels.AdditionalProperties[rh.tenantLabel] = id
+				rawRules.Groups[i].Rules[j] = r
+			case rules.AlertingRule:
+				if r.Labels.AdditionalProperties == nil {
+					r.Labels.AdditionalProperties = make(map[string]string)
+				}
+
+				r.Labels.AdditionalProperties[rh.tenantLabel] = id
+				rawRules.Groups[i].Rules[j] = r
+			}
+		}
+	}
```

Comment:
Do we need to enforce this label on the read path as well? Since the label is enforced during writes, we can just assume that it will be already present.

Reply:
The labels are not enforced anywhere in the write path currently. Also even if they wereb enforced on the write path, it wouldn't be enough because there's no telling how rules made it into a particular rule backend. In other words, we cannot be sure that the write path was even used at all to configure the backend. Perhaps the backend is read only and was configured using a static file outside of the workflow of making a PUT request to Observatorium, so there's no guarantee that the rules in the file have the right tenant ID. Finally, even though we proxy the tenant name in all requests to the backend, that's not good enough for the backend to enforce labels as the labels must include the tenant ID, not the tenant name.